sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+
+    @Nullable abstract ConnectionConfiguration getConnectionConfiguration();
+    @Nullable abstract String getQuery();
+    @Nullable abstract String getScrollKeepalive();
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration connectionConfiguration);
+      abstract Builder setQuery(String query);
+      abstract Builder setScrollKeepalive(String scrollKeepalive);
+      abstract Read build();
+    }
+
+    public Read withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "ElasticsearchIO.read()"
+          + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Read withQuery(String query) {
+      checkArgument(query != null, "ElasticsearchIO.read().withQuery(query) called with null "
+          + "query");
+      return builder().setQuery(query).build();
+    }
+
+    public Read withScrollKeepalive(String scrollKeepalive) {
+      checkArgument(scrollKeepalive != null, "ElasticsearchIO.read().withScrollKeepalive"
+          + "(keepalive) called with null keepalive");
+      return builder().setScrollKeepalive(scrollKeepalive).build();
+    }
+
+    @Override
+    public PCollection<String> apply(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(
+          new BoundedElasticsearchSource(this, null)
+      ));
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      checkState(getConnectionConfiguration() != null, "ElasticsearchIO.read() requires a "
+          + "connection configuration to be set via withConnectionConfiguration(configuration)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("query", getQuery()));
+      getConnectionConfiguration().populateDisplayData(builder);
+    }
+
+  }
+
+  /**
+   * A {@link BoundedSource} reading from Elasticsearch.
+   */
+  protected static class BoundedElasticsearchSource extends BoundedSource<String> {
+
+    private ElasticsearchIO.Read spec;
+    @Nullable
+    private final String shardPreference;
+
+    protected BoundedElasticsearchSource(Read spec, String shardPreference) {
+      this.spec = spec;
+      this.shardPreference = shardPreference;
+    }
+
+    @Override
+    public List<? extends BoundedSource<String>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                                  PipelineOptions options)
+        throws Exception {
+      List<BoundedElasticsearchSource> sources = new ArrayList<>();
+
+      // we split per shard
+      // unfortunately, Elasticsearch 2.x doesn't provide a way to do parallel reads on a single
+      // shard
```

Comment:
Explain a bit more about splitting here - e.g. what are "primary" shards? Perhaps link to the official docs about this?